### ISSUE TITLE
Support primtorch view ops in functionalization

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#83923]
I was starting to look at failing inductor models with `functorch.config.use_functionalize=True` turned on, and one failure that I noticed is:
```
// run this
python benchmarks/timm_models.py --float32 -dcuda --no-skip --training --inductor --use-eval-mode --only=mobilevit_s

// output
    return forward_call(*input, **kwargs)
  File "<eval_with_key>.4", line 394, in forward
    broadcast_in_dim_default = torch.ops.prims.broadcast_in_dim.default(var_default, [256, 256, 1], [0, 1]);  var_default = None
  File "/scratch/hirsheybar/work/benchmark/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: !schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED at "/scratch/hirsheybar/work/benchmark/pytorch/aten/src/ATen/FunctionalizeFallbackKernel.cpp":30, please report a bug to PyTorch. mutating and aliasing ops should all have codegen'd kernels
```

This error didn't show up in my AOT + eager tests, probably because the inductor backend is choosing to run extra primtorch decompositions. This causes functionalization to see `prims.broadcast_in_dim`, which is a new "view" op that it doesn't know how to handle.

For what it's worth - the code above fails with `use_functionalize=False` turned off today, too, because of some issue's in dynamo' `normalize_ir()` code. So this technically isn't a regression, although it seems like time would be better spent fixing the problem with functionalization instead of fixing dynamo's normalization logic.

Two potential ways to handle this are:
(1) write a custom functionalization kernel for `prims.broadcast_in_dims.default`. We'd need to expose the right C++ API's to python in order to do this
(2) Beef up the functionalization boxed fallback to handle views. This would be very useful, but it's a bit unclear how this would work: for every primtorch view, the boxed fallback needs to know how to map it to a "view inverse" function.

**Comment by ezyang** [comment]
I think the fastest path is to just port `prims.broadcast_in_dims` to C++, set it up as a proper view op.

**Comment by bdhirsh** [comment]
that probably is easiest - and assuming that we don't add view ops to prim torch too often, then hopefully won't require too much maintenance.

I just checked and we have 12 prim torch view ops though, so it will be a fair amount of boilerplate

It would be nice if we can just redirect them to the functionalization kernels for existing aten views, but unfortunately the names / overload names to match up 1-1 (but maybe it's worth getting that to work anyway since that should cover > 50% of the views):
```
imag(Tensor self) -> Tensor
real(Tensor self) -> Tensor
as_strided(Tensor(a!) a, int[] size, int[] stride, int storage_offset) -> Tensor(a!)
broadcast_in_dim(Tensor(a) a, SymInt[] shape, int[] broadcast_dimensions) -> Tensor(a)
collapse_view(Tensor(a) a, int start, int end) -> Tensor(a)
conj(Tensor(a) a) -> Tensor(a)
slice(Tensor(a) a, int[] start_indices, int[] limit_indices, int[]? strides=None) -> Tensor(a)
slice_in_dim(Tensor(a) a, int start_index, int limit_index, int stride=1, int axis=0) -> Tensor(a)
split_dim(Tensor(a) a, int dim, int outer_length) -> Tensor(a)
squeeze(Tensor(a) a, int[] dimensions) -> Tensor(a)
transpose(Tensor(a) a, int[] permutation) -> Tensor(a)
view_of(Tensor(a) a) -> Tensor
```

**Comment by ezyang** [comment]
TBH, I think it was a mistake to "force" all the prim ops to be named prim, and if there's an exact aten op that matches, we should have just used the "aten" op as the prim.

**Comment by Chillee** [comment]
@bdhirsh I think there are other options here.  For example, we could also perform decompositions after functionalization.

**Comment by Chillee** [comment]
Actually, if we just functionalized while doing the original tracing, wouldn't that also fix the issue?

**Comment by bdhirsh** [comment]
@Chillee yep I should totally have listed that as an option - doing functionalization first before decomps would help.

One thing I was worried about was that you might have a decomp that decomposes into mutating / view ops, which would mean that you're forced to functionalize afterwards anyway. I guess we could enforce that prim torch decomps that inductor opts into never do that, but it feels like of error prone.

**Comment by Chillee** [comment]
@bdhirsh I think we can just trace once without decompositions, and do the "decomposition" tracing in the trace we do with functionalization.

**Comment by bdhirsh** [comment]
> I think we can just trace once without decompositions, and do the "decomposition" tracing in the trace we do with functionalization.

Hmm, how does that alleviate this issue?

> you might have a decomp that decomposes into mutating / view ops, which would mean that you're forced to functionalize afterwards anyway

if you run `make_fx(functionalize(foo), decomps=DECOMP_TABLE)(inputs...)`, all of the decompositions will run "underneath" functionalization.
